### PR TITLE
AO3-6551 Correct skin CSS column type

### PR DESCRIPTION
**Pull request by leo60228**
# Pull Request Checklist

* [x] Have you read ["How to write the perfect pull request"](https://github.blog/2015-01-21-how-to-write-the-perfect-pull-request/)?
* [x] Have you read the [contributing guidelines](https://github.com/otwcode/otwarchive/blob/master/CONTRIBUTING.md)?
* [x] Have you added [tests for any changed functionality](https://github.com/otwcode/otwarchive/wiki/Automated-Testing)?
* [x] Have you added the [Jira](https://otwarchive.atlassian.net) issue number
  as the *first* thing in your pull request title (e.g. `AO3-1234 Fix thing`)
* [ ] Do you have fewer than 5 pull requests already open? If not, please wait
  until they are reviewed and merged before creating new pull requests.

## Issue

https://otwarchive.atlassian.net/browse/AO3-6551

## Purpose

Despite the intended limit being `CONTENT_MAX` (510,000 characters), attempting to create a skin with more than 65,535 characters of CSS would hit a 500 error as a result of the wrong SQL column type having been used. This adds a migration correcting it to `longtext`, placing it in line with other columns intended to have a limit of `CONTENT_MAX`.

Note that the JIRA ticket mentions having a JavaScript character counter for skins as a nice-to-have. ~~I am unsure how to implement this, and it seems separate from the actual issue.~~ This has been added as well.

## Testing Instructions

Try creating a very long skin, such as by copy/pasting [this contrived example](https://gist.github.com/leo60228/acf0d3a2a13695b5f3f8f80e15272355).

## References

Are there other relevant issues/pull requests/mailing list discussions?

## Credit

What name and pronouns should we use to credit you in the [Archive of Our Own's Release Notes](https://archiveofourown.org/admin_posts?tag=1)?

If you have a Jira account, please include the same name in the "Full name"
field on your Jira profile, so we can assign you the issues you're working on.

*Please note that if you do not fill in this section, we will use your GitHub account name and
they/them pronouns.*

### CODE REVIEW

**Comment by leo60228** [comment]
The test failure seems to be spurious? It's not related to anything that was changed.

**Comment by sarken** [comment]
Hi, leo60228!

Thank you so much for this pull request. Someone will be along to review it soon, but in the meantime, I've restarted the failed tests, which indeed appear to be some of our [known intermittently-failing tests](https://github.com/otwcode/otwarchive/wiki/Automated-Testing#user-content-Known_Test_Failures) and not related to your changes.

I've also updated the Jira issue status to In Review, so no one will mistakenly create a duplicate pull request. If you'd like the ability to comment on, assign, and transition issues in the future, you're welcome to create a Jira account! You can just reply here with the account name and we'll set up the permissions for you. (It makes things a bit easier for us on the organizational side if the Full Name on your Jira account either closely matches the name you'd like us to credit in the release notes or includes it in parentheses, e.g. "Nickname (PREFERRED NAME).")

Thanks again for contributing! If you have any questions, you can contact us at otw-coders@transformativeworks.org.

**Comment by weeklies** [comment]
The work form uses a [JS character counter](https://github.com/otwcode/otwarchive/blob/280200632a1248bc008e4c9586a6f8304ec6c97b/app/views/works/_standard_form.html.erb#L324-L333), you may want to try applying it for the CSS field.

**Comment by leo60228** [comment]
> I've also updated the Jira issue status to In Review, so no one will mistakenly create a duplicate pull request. If you'd like the ability to comment on, assign, and transition issues in the future, you're welcome to create a Jira account! You can just reply here with the account name and we'll set up the permissions for you. (It makes things a bit easier for us on the organizational side if the Full Name on your Jira account either closely matches the name you'd like us to credit in the release notes or includes it in parentheses, e.g. "Nickname (PREFERRED NAME).")

@sarken I've created an account with the name leo60228 (same name and email as GitHub).

> The work form uses a [JS character counter](https://github.com/otwcode/otwarchive/blob/280200632a1248bc008e4c9586a6f8304ec6c97b/app/views/works/_standard_form.html.erb#L324-L333), you may want to try applying it for the CSS field.

@weeklies Thanks, I've done that.

File: db/migrate/20230811004337_change_css_type.rb [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+          --max-load Threads_running=15 --critical-load Threads_running=100 \\
+          --set-vars innodb_lock_wait_timeout=2 --alter-foreign-keys-method=auto \\
+          --execute
+
+        Table Deletion Command:
+
+        DROP TABLE IF EXISTS `#{database}`.`_skins_old`;
+      PTOSC
+    else
+      change_column :skins, :css, "text"
```
Comment (sarken) [inline db/migrate/20230811004337_change_css_type.rb:46]: I think we can use `:text` here -- it only needs to be a string if it's a database-specific type.

File: db/migrate/20230811004337_change_css_type.rb [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+          --max-load Threads_running=15 --critical-load Threads_running=100 \\
+          --set-vars innodb_lock_wait_timeout=2 --alter-foreign-keys-method=auto \\
+          --execute
+
+        Table Deletion Command:
+
+        DROP TABLE IF EXISTS `#{database}`.`_skins_old`;
+      PTOSC
+    else
+      change_column :skins, :css, "longtext"
```
Comment (sarken) [inline db/migrate/20230811004337_change_css_type.rb:22]: A question for the peanut gallery: how do we feel about using database-specific types? The [documentation](https://api.rubyonrails.org/classes/ActiveRecord/ConnectionAdapters/SchemaStatements.html#method-i-add_column) says it "should usually be avoided," so do we have a strong preference for specifying `:text, limit: 4294967295` instead?

**Review by sarken — CHANGES_REQUESTED** [review]
Thank you so much for working on this!

File: app/views/skins/_form.html.erb [hunk 3]
```diff
@@ -31,7 +31,16 @@
 <fieldset>
   <legend><%= ts('CSS') %></legend>
   <h3 class="heading"><%= f.label :css, ts('CSS') %> <%= link_to_help 'skins-creating' %></h3>
-  <p><%= f.text_area :css, :cols => 70, :class => 'large' %></p>
+  <p>
+    <%= f.text_area :css, :cols => 70, :class => 'large observe_textlength' %>
+    <%= live_validation_for_field("skin_css",
+      maximum_length: ArchiveConfig.CONTENT_MAX_DISPLAYED,
+      tooLongMessage: ts("We salute your ambition! But sadly the content must be less than %{max} characters long.",
```
[the remark this answers is not in the document]
Comment (leo60228) [reply app/views/skins/_form.html.erb]: Done.

File: db/migrate/20230811004337_change_css_type.rb [hunk 4]
```diff
@@ -0,0 +1,49 @@
+class ChangeCssType < ActiveRecord::Migration[6.1]
+  def up
+    if Rails.env.staging? || Rails.env.production?
+      database = Kudo.connection.current_database
```
[the remark this answers is not in the document]
Comment (leo60228) [reply db/migrate/20230811004337_change_css_type.rb]: Done. I was looking at another migration for reference and assumed `Kudo` was the name of a helper class....

File: db/migrate/20230811004337_change_css_type.rb (hunk 1 above)
Comment (leo60228) [reply db/migrate/20230811004337_change_css_type.rb]: Done.

File: db/migrate/20230811004337_change_css_type.rb (hunk 2 above)
Comment (leo60228) [reply db/migrate/20230811004337_change_css_type.rb]: I changed it, if only for consistency with `down` now using `:text`.

**Comment by leo60228** [comment]
This seems to be another spurious failure? It's not one listed on the wiki, though.

**Comment by leo60228** [comment]
It looks like tests failed after merging because #4669 updated the schema version to 2023_09_20_094945, while the migration wasn't updated, so it gets skipped (since ActiveRecord thinks it was already applied). Should I submit another PR to update the migration filename?

**Comment by sarken** [comment]
Whoops, our bad! Production and staging have `config.active_record.dump_schema_after_migration = false` set, which avoids that problem, but we forgot to set it in the test.rb config. 

The coder who did [AO3-6639](https://otwarchive.atlassian.net/browse/AO3-6639)/#4669 is in the process of taking care of that change now, so this shouldn't need any action on your part, but thank you for pointing it out!

**Pull request merged**